### PR TITLE
Cleanup of threading, client/server, RegisterLuaType, and filesystem stuff

### PR DESCRIPTION
**Pull request by Meisaka**
## Description
I know it might sound lazy of me, but verbose details are in the commit messages.
It's well worth reviewing on a commit by commit basis.

## Motivation and Context
- See #222 

## How Has This Been Tested?
built on Archlinux :dango: and run: tests, client and server

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by adam4813** [comment]
The spawning of so many threads instead of pooling on Linux is annoying and only on Linux. The other platforms appear to use thread pools of launch_async.

Everything looks good and is a good cleanup!

**Review by adam4813 — APPROVED** [review]

File: client/main.cpp [hunk 1]
```diff
@@ -189,12 +189,15 @@ int main(int argc, char* argv[]) {
...
 				if (splitArgs.size() >= 2) {
 					connection.Connect(splitArgs[1]);
-					tec::proto::UserLogin user_login;
-					user_login.set_username(splitArgs[0]);
-					user_login.set_password("");
-					tec::networking::MessageOut msg(tec::networking::LOGIN);
-					user_login.SerializeToZeroCopyStream(&msg);
-					connection.Send(msg);
+					std::string username = splitArgs[0];
+					connection.RegisterConnectFunc([&connection, username]() {
```
Comment (adam4813) [inline client/main.cpp:193]: I realize that RegisterConnectFunc naming maybe odd at this point in my career. Perhaps OnConnect would be more appropriate given that it is an event callback.

File: client/os.cpp [hunk 2]
```diff
@@ -115,7 +115,7 @@ bool OS::InitializeWindow(
 	std::string glsl_minor = glcx_glslver.substr(glcx_glslver.find('.', 0) + 1, 1);
 	if (glsl_major.at(0) < '3') {
 		l->critical(
-				"[OS] Initializing OpenGL failder, Shader version must be >= 3.30 : GL version : "
+				"[OS] Initializing OpenGL failed, Shader version must be >= 3.30 : GL version : "
```
Comment (adam4813) [inline client/os.cpp:118]: Wow double failder 🥇

File: common/simulation.cpp [hunk 3]
```diff
@@ -14,6 +14,30 @@ namespace tec {
 
 double UPDATE_RATE = 1.0 / 8.0; // 8 per second
 double TICKS_PER_SECOND = 60.0 * UPDATE_RATE;
+
+Simulation::Simulation() {
+	worker_thread = new std::thread([this]() {
+		std::unique_lock lock(worker_m);
+		while (this->worker_thread) {
+			worker_cv.wait_for(lock, std::chrono::milliseconds(1));
+			if(!lock.owns_lock()) {
```
Comment (adam4813) [inline common/simulation.cpp:23]: Where does it get unlocked?

Comment (Meisaka) [reply common/simulation.cpp:23]: the call to wait_for will unlock while waiting to be signaled, then take the lock again before returning, so the thread should be holding the lock when signaled.

Comment (adam4813) [reply common/simulation.cpp:23]: Oh so the thread sleeps for 1 ms? Should it just wait for the cv to be notified? If you need to watch dog shouldn't it be longer like seconds?

**Pull request merged**